src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -104,17 +129,73 @@ public void run() {
+    private Stream<Thing> getThingsAssociatedWithCoreDevice() {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .initialRetryInterval(Duration.ofSeconds(30)).maxRetryInterval(Duration.ofMinutes(3))
+                .maxAttempt(3)
+                .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class)).build();
+
+        try {
+            return RetryUtils.runWithRetry(retryConfig, iotAuthClient::getThingsAssociatedWithCoreDevice,
+                    "get-things-associated-with-core-device", logger);
+        } catch (AccessDeniedException e) {
+            logger.atDebug().log(
+                    "Access denied to list things associated with core device. Please make sure you "
+                    + "have setup the correct permissions.");
+        } catch (Exception e) {
+            logger.warn("Failed to get things associated to the core device. Retry will be schedule later");
+        }
+
+        return Stream.empty();
+    }
+
+    private Consumer<Thing> refresh() {
+        Set<String> alreadyRefreshed = new HashSet<>();
+
+       return (Thing thing) -> {
+           Set<String> certificateIds = thing.getAttachedCertificateIds().keySet();
+           // Update thing certificate attachments
+           certificateIds.stream()
+                   .map(certificateId -> new VerifyThingAttachedToCertificateDTO(thing, new Certificate(certificateId)))
+                   .forEach(this::refreshCertificateThingAttachment);
+
+           // Update the certificate validity for certificates that have not been updated
+           certificateIds.stream()
+                   .filter((certificateId) -> !alreadyRefreshed.contains(certificateId))
+                   .map((certificateId) -> new Pair<>(certificateId, pemStore.getPem(certificateId)))
+                   .filter(pair -> pair.getRight().isPresent())
+                   .forEach(pair -> {
+                       this.refreshCertificateValidity(pair.getRight().get());
+                       alreadyRefreshed.add(pair.getLeft());
+                   });
+       };
+    }
 
-        for (String certificateId : certificateIds) {
-            if (pemStore.exists(certificateId)) {
-                String pem = pemStore.getPem(certificateId).get();
-                verifyIotCertificate.apply(pem);
-            }
+    private void refreshCertificateThingAttachment(VerifyThingAttachedToCertificateDTO dto) {
+        VerifyThingAttachedToCertificate useCase = useCases.get(VerifyThingAttachedToCertificate.class);
+
+        try {
+            useCase.apply(dto);
+        } catch (AuthenticationException e) {
+            logger.warn("Failed to refresh certificate thing attachment {}", dto.getThing().getThingName(), e);
         }
     }
 
+    private void refreshCertificateValidity(String pem) {
+        VerifyIotCertificate useCase = useCases.get(VerifyIotCertificate.class);
+        useCase.apply(pem);
+    }
```

Comment:
should we also remove PEM from the filesystem? I see `VerifyIoTCertificate` usecase only updates metadata in the registry.

Reply:
^ for the certificates found invalid while _refresh_

Reply:
We could do that check the ones that we have that are not on the list of associated to the core and delete them.